### ISSUE TITLE
project name not support more than 2 components,  V2 registry API does not enforce this

### ISSUE THREAD

**Issue by yayuntian** [vmware/harbor#3162]
# docker version
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:20:36 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:21:56 2017
 OS/Arch:      linux/amd64
 Experimental: false

#Problem recurrence
1, created  "aaaa" project by harbor-ui
2, use docker push xxxx.com/aaaa/bb/2048:0.1  is ok
3, start replicate to another node，then go panic

check code found error occur in  Checker -> createProject -> validateProjectReq

```
func validateProjectReq(req projectReq) error {
	pn := req.ProjectName
	if isIllegalLength(req.ProjectName, projectNameMinLen, projectNameMaxLen) {
		return fmt.Errorf("Project name is illegal in length. (greater than 2 or less than 30)")
	}
	validProjectName := regexp.MustCompile(`^` + restrictedNameChars + `$`)
	legal := validProjectName.MatchString(pn)
	if !legal {
		return fmt.Errorf("project name is not in lower case or contains illegal characters")
	}
	return nil
}
```

also found get project name use func
```
func getProjectName(repository string) string {
	repository = strings.TrimSpace(repository)
	repository = strings.TrimRight(repository, "/")
	return repository[:strings.LastIndex(repository, "/")]
}
```

```
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [replication.go:73]: data: {PolicyID:35 Repo: Operation:transfer TagList:[]}
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [replication.go:92]: repo list: [aaaa/bbb/2048]
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [replication.go:124]: Creating job for repo: aaaa/bbb/2048, policy: 35
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [replication.go:129]: Send job to scheduler, job id: 1589
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017/09/06 03:53:11 #033[1;34m[router.go:829][D] | POST       | /api/jobs/replication                    | 9.402363ms       | match      | /api/jobs/replication                    | #033[0m
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [workerpool.go:137]: Trying to dispatch job: 1589
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [workerpool.go:62]: worker: 1, will handle job: 1589
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:61]: Job id: 1589, transiting from State: pending, to State: running
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:74]: Job id: 1589, no handler found for state:pending, skip
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:88]: Job id: 1589, transition succeeded, current state: running
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:97]: Job id: 1589, next state from handler: _continue
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:109]: Job id: 1589, Continue to state: initialize
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:61]: Job id: 1589, transiting from State: running, to State: initialize
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [transport.go:50]: 200 | GET http://registry:5000/v2/aaaa/bbb/2048/tags/list
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:88]: Job id: 1589, transition succeeded, current state: initialize
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:118]: Job id: 1589, next state from handler: check
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 2017-09-06T03:53:11Z [DEBUG] [statemachine.go:61]: Job id: 1589, transiting from State: initialize, to State: check
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: panic: runtime error: invalid memory address or nil pointer dereference
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x8ed5a7]
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: 
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: goroutine 52 [running]:
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: github.com/vmware/harbor/src/jobservice/replication.(*Checker).enter(0xc42000e0a0, 0xd76c313890a4feb8, 0xc4202dda70, 0x415387, 0xc4202dda90)
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: #011/root/go/src/github.com/vmware/harbor/src/jobservice/replication/transfer.go:195 +0x187
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: github.com/vmware/harbor/src/jobservice/replication.(*Checker).Enter(0xc42000e0a0, 0xc42015c6f0, 0xabc250, 0x5, 0xc42036a9b8)
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: #011/root/go/src/github.com/vmware/harbor/src/jobservice/replication/transfer.go:179 +0x32
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: github.com/vmware/harbor/src/jobservice/job.(*SM).EnterState(0xc4202f9c00, 0xabc250, 0x5, 0x2, 0x2, 0x0, 0x0)
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: #011/root/go/src/github.com/vmware/harbor/src/jobservice/job/statemachine.go:80 +0x2e5
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: github.com/vmware/harbor/src/jobservice/job.(*SM).Start(0xc4202f9c00, 0xabed66, 0x7)
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: #011/root/go/src/github.com/vmware/harbor/src/jobservice/job/statemachine.go:117 +0x20a
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: github.com/vmware/harbor/src/jobservice/job.(*Worker).handleRepJob(0xc420327980, 0x635)
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: #011/root/go/src/github.com/vmware/harbor/src/jobservice/job/workerpool.go:96 +0x3ff
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: github.com/vmware/harbor/src/jobservice/job.(*Worker).Start.func1(0xc420327980)
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: #011/root/go/src/github.com/vmware/harbor/src/jobservice/job/workerpool.go:63 +0x1e2
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: created by github.com/vmware/harbor/src/jobservice/job.(*Worker).Start
Sep  6 11:53:11 172.22.0.1 jobservice[2054]: #011/root/go/src/github.com/vmware/harbor/src/jobservice/job/workerpool.go:71 +0x3f

```

**Comment by yayuntian** [comment]
https://docs.docker.com/registry/spec/api/#overview  

```
Classically, repository names have always been two path components where each path component is 
less than 30 characters. The V2 registry API does not enforce this. The rules for a repository name are as follows:

1. A repository name is broken up into path components. A component of a repository name must 
be at least one lowercase, alpha-numeric characters, optionally separated by periods, 
dashes or underscores. More strictly, it must match the regular
 expression [a-z0-9]+(?:[._-][a-z0-9]+)*.

2. If a repository name has two or more path components, 
they must be separated by a forward slash (“/”).The total length of a repository name,
 including slashes, must be less than 256 characters.

3. These name requirements only apply to the registry API and should accept a superset of what is
 supported by other docker ecosystem components.

```

**Comment by ywk253100** [comment]
This is a known issue and has been fixed in version 1.2, please upgrade to the latest version.
